lib/Dialect/TritonGPU/Transforms/Pipeliner/PipelineExpander.cpp

```diff
@@ -222,6 +240,101 @@ static SetVector<Value> getNestedOperands(Operation *op) {
   return operands;
 }
 
+// void LoopPipelinerInternal::printLoopArgConflicts(Operation *consumer,
+//                                                   Value operand,
+//                                                   Operation *producer) {
+//   if (auto arg = dyn_cast<BlockArgument>(operand)) {
+//     std::optional<Value> yieldValue = getBlockArgYieldValueFromForLoop(arg);
+//     if (yieldValue) {
+//       Value yieldValue_ = *yieldValue;
+//       yieldValue_.
+//     }
+//   }
+// }
+
+DenseSet<Operation *> findRootDefiningOp(Operation *op,
```

Comment:
I thought what we had discussed was to have this logic separate from the transformation code?

Reply:
I see. Would it be acceptable to encapsulate this logic into a separate class and move it to a new source file? This way, we can maintain a clear separation between the transformation code and this logic.

Reply:
what part of this file do you need? Is it just the check to know if the schedule is valide?

Reply:
I am not sure if I get your question, but this PR does not introduce more checks. This PR provides detailed error message when distance = 1, per discussions with Pawel. `findRootDefiningOp` is to support the detailed error message.

I can get your point to separate checks from transformation. For this PR, how about I put related impl of `printSchedulingError` into an separate file such as `ErrorPrinter.cpp`?

Reply:
Trying to understand what is exactly needed here. Do we want a detailed error message for compiler developers to help them fix the issue, or just a message to the user that the compiler failed to fulfill a schedule? 

For the former, we may need a message tied to the transformation and the current implementation looks helpful.

For the latter we might need more simplified messaging.

Reply:
@htyu This PR addresses the first scenario you mentioned. It provides a detailed error message for compiler developers, which helps in identifying the exact line related to the issue discussed in #5172. You can see this demonstrated in the unit test included with the PR.

Reply:
I see. In that case it sounds ok for the diagnosis logics to be pipeliner-specific. 

@ThomasRaoux I may misunderstand your point. Would you like this logics to be in a separate file or to have a less pipeliner look?

Reply:
My point was that it would be good to do this outside of the expander rather. It doesn't look like it needs to reuse much so this could be implemented as a separate verification + diagnostic

Reply:
I agree that we should make this independent of the expander. Expander is supposed to stay close to the LLVM upstream implementation, and most likely replaced by it someday. The new error is more informative and actionable, which is great, but it would be great if we could make it independent from the expander. Perhaps a verification pass that we pass the schedule vector to?

Reply:
@pawelszczerbuk I see. Passing the schedule vector to a separate pass makes sense to me. Do you want to refactor the existing verification part in the expander to a new pass as well? Right now I am only adding an extra error reporting to the verification. My PR lacks some comments - it looks complicated but they just collect more information for the error message. There's no extra verification. 

Thank you all for the discussion! Anyhow, let me first improve my code quality: 1) adding more comments and 2) moving into a separate file. Then I will try to further separate them into a pass.

Reply:
Thanks @sfzhu93 ! Let's not modify the expander for now. If we'll do it, it should be separate PR that is in sync with change in the upstream LLVM expander. It can come later.

Reply:
@pawelszczerbuk Thanks! I have cleaned up the code and I have minimum change to the expander. The PR is not yet finished - to make the error more informative, the error reporter will likely depend on the schedule of each Op as well.

 I can further make it into a separate pass if you feel it is necessary.